solr/core/src/java/org/apache/solr/search/facet/FacetRangeProcessor.java

```diff
@@ -539,7 +617,116 @@ private  SimpleOrderedMap getRangeCountsIndexed() throws IOException {
 
   private Query[] filters;
   private DocSet[] intersections;
+  
   private void rangeStats(Range range, int slot, boolean hasSubFacets) throws IOException {
+    Query rangeQ = getRangeQuery(range);
+    // TODO: specialize count only
+    DocSet intersection = fcontext.searcher.getDocSet(rangeQ, fcontext.base);
+    if (hasSubFacets) {
+      filters[slot] = rangeQ;
+      intersections[slot] = intersection;  // save for later  // TODO: only save if number of slots is small enough?
+    }
+    int num = collect(intersection, slot, slotNum -> { return new SlotAcc.SlotContext(rangeQ); });
+    countAcc.incrementCount(slot, num); // TODO: roll this into collect()
+  }
+
+  private void dvRangeStats(List<Range> rangeList, List<Range> otherList, boolean hasSubFacets) throws IOException {
+    List<Range> allRanges = new ArrayList<>(rangeList.size() + otherList.size());
+    allRanges.addAll(rangeList); // first rangeList items are sorted by low
+    allRanges.addAll(otherList);
+    
+    final FieldType ft = sf.getType();
+    final String fieldName = sf.getName();
+    final NumberType numericType = ft.getNumberType();
+    if (numericType == null) {
+      throw new IllegalStateException();
+    }
+
+    final IndexReader indexReader = fcontext.searcher.getIndexReader();
+    final int maxDoc = indexReader.maxDoc();
+    final Iterator<LeafReaderContext> ctxIt = indexReader.leaves().iterator();
+    LeafReaderContext ctx = null;
+    NumericDocValues longs = null;
+    DocSetBuilder[] setBuilders = hasSubFacets ? new DocSetBuilder[intersections.length] : null;
+    
+    for (DocIterator docsIt = fcontext.base.iterator(); docsIt.hasNext(); ) {
+      final int doc = docsIt.nextDoc();
+      if (ctx == null || doc >= ctx.docBase + ctx.reader().maxDoc()) {
+        do {
+          ctx = ctxIt.next();
+        } while (ctx == null || doc >= ctx.docBase + ctx.reader().maxDoc());
+        assert doc >= ctx.docBase;
+        switch (numericType) {
+          case LONG:
+          case DATE:
+          case INTEGER:
+            longs = DocValues.getNumeric(ctx.reader(), fieldName);
+            break;
+          case FLOAT:
```

Comment:
The handling of Float, Double is sub-optimal since a new object is created for every field. We can just create one Object and reuse it.

Reply:
It only creates the new object (assuming you mean the `FilterNumericDocValues` object) when the leaf reader context needs to change. So it shouldn't be many instances. Plus, we're required create a new `NumericDocValues` instance regardless given the new leaf reader.  I think that's acceptable?